Scrollable span details

stoeffel opened the pull request:
fixes #17

Commit:
Use a scrollable container for span details

Commit:
scroll span details with 1 and 2

coming next: make the right pane focusable and use the standard
keybindings for scrolling

stoeffel committed:
Add joinable borders to breakdown page

stoeffel commented on nri-log-explorer/src/Main.hs:
```diff
@@ -628,6 +630,7 @@ view model =
             viewContents page,
             viewKey page (clipboardCommand model)
           ]
+          |> Brick.joinBorders
```
This needs to be at the top for some reason.

stoeffel commented on nri-log-explorer/src/Main.hs:
```diff
@@ -907,77 +913,81 @@ viewSpanBreakdownWithMatch matching span =
-        Platform.Succeeded -> viewDetail "result" "succeeded"
-        Platform.Failed -> viewDetail "result" "failed"
-        Platform.FailedWith exception ->
+  Brick.viewport
```
I find it interesting that you added this inside `viewSpanDetails` where as I would have added it outside. I don't think it matters, but it's interesting how preferences are different.
I would have added it outside, because I feel like it's not part of rendering span details.

stoeffel replied:
Also because you can write
```hs
viewSpanDetails span
  |> Brick.viewport SpanDetail Vertical
```
without having so much nesting inside here.

juanedi replied:
> I find it interesting that you added this inside viewSpanDetails where as I would have added it outside. I don't think it matters, but it's interesting how preferences are different.

FWIW I also think that having it outside is better (reasoning: that feels more like "how it fits into the parent's layout"). I suppose that at the time I was so new to the code I didn't really understand what I was doing well enough to make that type of decision :-P

juanedi committed:
Merge remote-tracking branch 'origin/trunk' into scrollable-span-details

juanedi committed:
keep track of focused element in breakdown page

juanedi committed:
keybindings to switch currently focused pane

juanedi committed:
ormolu

juanedi committed:
allow focusing on the details by pressing enter

juanedi committed:
i got the focus highlight backwards!

juanedi committed:
adding keybinding hints at the bottom

hints for how to:
  - toggle the focused pane
  - scroll when focusing on the span details (TODO: we still need to map to the right keys!)

juanedi commented:
Status update after a brief pairing with @marcellerusu.

Done:
  - keeping track of currently focused pane and toggling with tab/backtab
  - ugly visual hint to be able to distinguish the currently focused pane (just went with whatever first occurred to test the logic)
  - keybinding hints at the bottom

To-dos that come to mind:
  - [x] bind scrolling to up/down and maybe j/k
  - [x] only scroll span details if we are focusing on that pane
  - [ ] visually pleasing hint of which is the current pane

juanedi commented:
I'll try to keep up the momentum and find the time to implement those this week.

stoeffel reviewed (commented):
Looks good so far!

juanedi committed:
Only scroll list widget if we are focusing on the left pane

juanedi committed:
scroll span details with j/k too

juanedi committed:
only scroll span details while focusing on the span details!

juanedi committed:
used coloured borders to indicate the focused pane

juanedi commented:
Update: coloured borders are working but before merging I would still like to find a way to make the visual hint more clear. Since we always keep the center border highlighted, the only thing that changes is the side borders which are a bit hard to notice.

https://user-images.githubusercontent.com/753421/142001438-c35b4e47-6ebc-4bab-bd28-e74ff3f0f445.mov

stoeffel committed:
Proposal for focus styles

stoeffel commented:
@juanedi I pushed a proposal for some focus styling. Code is ugly, but shows some ideas on how we can improve styling

juanedi commented:
Ohhhh I love it!!! It didn't occur to me that we could even do that 😅

juanedi commented on nri-log-explorer/src/Main.hs:
```diff
@@ -971,10 +985,13 @@ viewSpanBreakdown spans =
               else identity
       )
-      True
+      (focus == FocusOnSpanList)
```
It was hard for me to understand that `hasFocus` here refers to whether the current element is being focus, and that it was not referring also to whether the list widget was active. How would you feel about a rename in the lines of...:
- `hasFocus` -> `isSpanFocused`
- `focus` -> `focusedPane`

juanedi replied:
Did that as part of https://github.com/NoRedInk/haskell-libraries/pull/19/commits/6402aac7b4397a3f47d20d917cc7147c42602b4c, along with other renames

juanedi commented on nri-log-explorer/src/Main.hs:
```diff
@@ -1275,7 +1292,9 @@ attrMap =
-      )
+      ),
+      ("dim-unfocused", Vty.withStyle Vty.defAttr Vty.Attributes.dim),
+      ("focused", Vty.withStyle Vty.defAttr Vty.Attributes.standout)
```
Cool! It didn't even occur to me that `Vty` would provide these styles. I guess I still don't know exactly what `Vty` is in charge of vs what falls under Brick's responsibilities 😅

juanedi committed:
refactor in span breakdown page view

trying to consolidate conditional styles based on the current pane in a single
place, separate from the layout.

juanedi committed:
:art: rename type for currently focused pane

  - SpanBreakdownPageFocus -> SpanBreakdownFocusedPane
  - FocusOnSpanList -> SpanList
  - FocusOnSpanDetails -> SpanDetails

juanedi committed:
TO-DO -> TO-DONE

juanedi commented:
🍐 'ed with @marcellerusu on some refactoring/renames. We didn't make any dramatic changes to the code. Mostly trying to consolidate all the conditional styling in one place.

@stoeffel this is ready for official review as far as we are concerned 🤘

stoeffel committed:
Slightly simplify side by side view

stoeffel commented:
reviewed together with @juanedi ! thanks
